File: src/test/java/cube8540/oauth/authentication/users/application/UserAutoApprovalScopeHandlerTest.java
package cube8540.oauth.authentication.users.application;

import cube8540.oauth.authentication.oauth.security.OAuth2ClientDetails;
import cube8540.oauth.authentication.users.domain.ApprovalAuthority;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.security.Principal;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.ALL_APPROVAL_SCOPES;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.AUTHORITIES_A;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.CLIENT_A_APPROVAL_AUTHORITIES;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.CLIENT_A_ID;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.RAW_USERNAME;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.REQUEST_SCOPES;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeAuthentication;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeClientDetails;
import static cube8540.oauth.authentication.users.application.UserApplicationTestHelper.makeComposeApprovalAuthorities;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@DisplayName("유저 자동 승인 스코프 헨들러 테스트")
public class UserAutoApprovalScopeHandlerTest {

    @Test
    @DisplayName("자동 승인 스코프 필터링")
    void filterRequiredApprovalScopes() {
        UserApprovalAuthorityService approvalService = mock(UserApprovalAuthorityService.class);
        Principal authentication = makeAuthentication(RAW_USERNAME);
        OAuth2ClientDetails clientDetails = makeClientDetails(CLIENT_A_ID);
        Set<ApprovalAuthority> userApprovalAuthorities = makeComposeApprovalAuthorities();
        UserAutoApprovalScopeHandler handler = new UserAutoApprovalScopeHandler(approvalService);

        when(approvalService.getApprovalAuthorities(RAW_USERNAME)).thenReturn(userApprovalAuthorities);

        Set<String> results = handler.filterRequiredPermissionScopes(authentication, clientDetails, REQUEST_SCOPES);
        assertEquals(subtract(REQUEST_SCOPES, CLIENT_A_APPROVAL_AUTHORITIES), results);
    }

    @Test
    @DisplayName("모든 스코프가 자동 승인 일때")
    void filterRequiredApprovalScopeWhenAlreadyAllScopeIsAutoApproval() {
        UserApprovalAuthorityService approvalService = mock(UserApprovalAuthorityService.class);
        Principal authentication = makeAuthentication(RAW_USERNAME);
        OAuth2ClientDetails clientDetails = makeClientDetails(CLIENT_A_ID);
        Set<ApprovalAuthority> userApprovalAuthorities = makeComposeApprovalAuthorities();
        UserAutoApprovalScopeHandler handler = new UserAutoApprovalScopeHandler(approvalService);

        when(approvalService.getApprovalAuthorities(RAW_USERNAME)).thenReturn(userApprovalAuthorities);

        Set<String> results = handler.filterRequiredPermissionScopes(authentication, clientDetails, ALL_APPROVAL_SCOPES);
        assertEquals(Collections.emptySet(), results);
    }

    @Test
    @DisplayName("자동 승인 스코프 저장")
    void storeAutoApprovalScope() {
        UserApprovalAuthorityService approvalService = mock(UserApprovalAuthorityService.class);
        Principal authentication = makeAuthentication(RAW_USERNAME);
        OAuth2ClientDetails clientDetails = makeClientDetails(CLIENT_A_ID);
        UserAutoApprovalScopeHandler handler = new UserAutoApprovalScopeHandler(approvalService);

        handler.storeAutoApprovalScopes(authentication, clientDetails, CLIENT_A_APPROVAL_AUTHORITIES);
        verify(approvalService, times(1)).grantApprovalAuthorities(RAW_USERNAME, AUTHORITIES_A);
    }

    private Set<String> subtract(Set<String> a, Set<String> b) {
        Set<String> newA = new HashSet<>(a);

        newA.removeAll(b);
        return newA;
    }

}

File: src/test/kotlin/cube8540/oauth/authentication/user/application/UserAutoApprovalScopeHandlerTest.kt
package cube8540.oauth.authentication.user.application

import cube8540.oauth.authentication.oauth.security.OAuth2ClientDetails
import cube8540.oauth.authentication.users.application.UserApprovalAuthorityService
import cube8540.oauth.authentication.users.application.UserAutoApprovalScopeHandler
import cube8540.oauth.authentication.users.domain.ApprovalAuthority
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.security.Principal

class UserAutoApprovalScopeHandlerTest {

    private val approvalService: UserApprovalAuthorityService = mockk()
    private val handler = UserAutoApprovalScopeHandler(approvalService)

    @Test
    fun `filtering required approval scope`() {
        val requestApprovalScope = setOf("TEST-1", "TEST-2", "TEST-3", "TEST-4", "TEST-5")
        val approvalScopeInClientA = listOf("TEST-1", "TEST-2", "TEST-3")
        val approvalScopeInClientB = listOf("TEST-2", "TEST-3", "TEST-4")
        val approvalScopeInClientC = listOf("TEST-3", "TEST-4", "TEST-5")
        val clientDetails: OAuth2ClientDetails = mockk {
            every { clientId } returns "clientA"
        }
        val principal: Principal = mockk {
            every { name } returns "username"
        }
        val userApprovalScopes = approvalScopeInClientA.map { ApprovalAuthority("clientA", it) } +
                approvalScopeInClientB.map { ApprovalAuthority("clientB", it) } +
                approvalScopeInClientC.map { ApprovalAuthority("clientC", it) }

        every { approvalService.getApprovalAuthorities("username") } returns userApprovalScopes

        val results = handler.filterRequiredPermissionScopes(principal, clientDetails, requestApprovalScope)
        assertThat(results).isEqualTo(requestApprovalScope.subtract(approvalScopeInClientA))
    }

    @Test
    fun `save auto approval scope`() {
        val storeApprovalCaptor = slot<Collection<ApprovalAuthority>>()
        val requestApprovalScope = setOf("TEST-1", "TEST-2", "TEST-3")
        val principal: Principal = mockk {
            every { name } returns "username"
        }
        val clientDetails: OAuth2ClientDetails = mockk {
            every { clientId } returns "clientId"
        }

        every { approvalService.grantApprovalAuthorities(eq("username"), capture(storeApprovalCaptor)) } returns mockk()

        handler.storeAutoApprovalScopes(principal, clientDetails, requestApprovalScope)
        assertThat(storeApprovalCaptor.isCaptured).isTrue
        assertThat(storeApprovalCaptor.captured).isEqualTo(
            requestApprovalScope.map { ApprovalAuthority("clientId", it) }.toSet()
        )
    }
}